react-firebase-email-verification.md

```diff
@@ -0,0 +1,412 @@
+    });
+
+  ...
+}
+
+export default Firebase;
+{{< /highlight >}}
+
+You can inline this URL, but also put it into your *.env* file(s). I prefer environmental variables for development (*.env.development*) and production (*.env.production*). The development environment receives the localhost URL:
+
+{{< highlight javascript "hl_lines=3" >}}
+...
+
+REACT_APP_CONFIRMATION_EMAIL_REDIRECT=http://localhost:3000
+{{< /highlight >}}
+
+And the production environment receives an actual domain:
+
+{{< highlight javascript "hl_lines=3" >}}
+...
+
+REACT_APP_CONFIRMATION_EMAIL_REDIRECT=https://mydomain.com
+{{< /highlight >}}
+
+That's all we need to do for the API. The best place to guide users through the email verification is during email and password sign-up:
+
+{{< highlight javascript "hl_lines=19 20 21" >}}
+...
+
+class SignUpFormBase extends Component {
+  ...
+
+  onSubmit = event => {
+    ...
+
+    this.props.firebase
+      .doCreateUserWithEmailAndPassword(email, passwordOne)
+      .then(authUser => {
+        // Create a user in your Firebase realtime database
+        return this.props.firebase.user(authUser.user.uid).set({
+          username,
+          email,
+          roles,
+        });
+      })
+      .then(() => {
+        return this.props.firebase.doSendEmailVerification();
+      })
+      .then(() => {
+        this.setState({ ...INITIAL_STATE });
+        this.props.history.push(ROUTES.HOME);
+      })
+      .catch(error => {
+        ...
+      });
+
+    event.preventDefault();
+  };
+
+  ...
+}
+
+...
+{{< /highlight >}}
+
+Users will receive a verification email when they register for your application. To find out if a user has a verified email, you can retrieve this information from the authenticated user in your Firebase class:
+
+{{< highlight javascript "hl_lines=25 26" >}}
+...
+
+class Firebase {
+  ...
+
+  // *** Merge Auth and DB User API *** //
+
+  onAuthUserListener = (next, fallback) =>
+    this.auth.onAuthStateChanged(authUser => {
+      if (authUser) {
+        this.user(authUser.uid)
+          .once('value')
+          .then(snapshot => {
+            const dbUser = snapshot.val();
+
+            // default empty roles
+            if (!dbUser.roles) {
+              dbUser.roles = [];
+            }
+
+            // merge auth and db user
+            authUser = {
+              uid: authUser.uid,
+              email: authUser.email,
+              emailVerified: authUser.emailVerified,
+              providerData: authUser.providerData,
+              ...dbUser,
+            };
+
+            next(authUser);
+          });
+      } else {
+        fallback();
+      }
+    });
+
+    ...
+}
+
+export default Firebase;
+{{< /highlight >}}
+
+To protect your routes from users who have no verified email address? We will do it with a new higher-order component in *src/components/Session/withEmailVerification.js* that has access to Firebase and the authenticated user:
```

Comment:
"To protect your routes from users who have no verified email address?" <- should this be a question?